File: src/render/overlay.ts
import { clamp01, lerp } from '../lib/math';
import type { FrameMapper } from '../tracking/frameMapper';
import { FINGER_BONES, LM, PALM_BONES, type FingerName } from '../tracking/landmarks';
import type { HandPointer } from '../tracking/pointers';
import type { Tool } from '../types';
import { withAlpha } from './color';

const TAU = Math.PI * 2;
const CHARGE_MS = 170;

export interface OverlayOptions {
  showSkeleton: boolean;
  tool: Tool;
  color: string;
  /** Current brush size in board units. */
  brush: number;
  /** performance.now() for time-based cues. */
  now: number;
  trace: string;
  dim: string;
  paper: string;
}

const OTHER_FINGERS: readonly FingerName[] = ['thumb', 'middle', 'ring', 'pinky'];

/**
 * Draws everything that is not ink: the tracked hand skeleton and the nib.
 *
 * The nib is deliberately more than a dot. Air-drawing has no physical
 * feedback, so the cursor has to answer three questions at a glance: where the
 * pen is, how thick the line will be (the ring is the true stroke width), and
 * how close the gesture is to engaging (the arc). Without that last one, a
 * gesture that nearly fires just feels broken.
 */
export class OverlayRenderer {
  private readonly charges = new Map<number, number>();

  render(
    ctx: CanvasRenderingContext2D,
    mapper: FrameMapper,
    pointers: readonly HandPointer[],
    options: OverlayOptions,
    width: number,
    height: number,
  ): void {
    ctx.clearRect(0, 0, width, height);
    if (!mapper.ready) return;

    for (const pointer of pointers) {
      if (pointer.justPressed) this.charges.set(pointer.id, options.now);
      if (options.showSkeleton) this.drawSkeleton(ctx, mapper, pointer, options);
      if (!pointer.isDown) this.drawTrail(ctx, mapper, pointer, options);
    }

    // Nibs last so they sit above every skeleton when hands overlap.
    for (const pointer of pointers) this.drawNib(ctx, mapper, pointer, options);
  }

  private drawSkeleton(
    ctx: CanvasRenderingContext2D,
    mapper: FrameMapper,
    pointer: HandPointer,
    options: OverlayOptions,
  ): void {
    const project = (index: number): { x: number; y: number } => {
      const lm = pointer.landmarks[index];
      return mapper.normalizedToStage(lm.x, lm.y);
    };

    // Step back while drawing so the skeleton never competes with the stroke.
    const fade = pointer.isDown ? 0.45 : 1;

    ctx.save();
    ctx.lineCap = 'round';

    const bone = (
      pairs: ReadonlyArray<readonly [number, number]>,
      alpha: number,
      lineWidth: number,
      color: string,
    ): void => {
      ctx.strokeStyle = withAlpha(color, alpha * fade);
      ctx.lineWidth = lineWidth;
      ctx.beginPath();
      for (const [a, b] of pairs) {
        const from = project(a);
        const to = project(b);
        ctx.moveTo(from.x, from.y);
        ctx.lineTo(to.x, to.y);
      }
      ctx.stroke();
    };

    bone(PALM_BONES, 0.28, 1.25, options.trace);
    for (const finger of OTHER_FINGERS) bone(FINGER_BONES[finger], 0.28, 1.25, options.trace);
    // The index chain is the one that matters — it carries the nib.
    bone(FINGER_BONES.index, 0.85, 1.75, options.trace);

    ctx.fillStyle = withAlpha(options.trace, 0.5 * fade);
    for (let i = 0; i < pointer.landmarks.length; i += 1) {
      if (i === LM.INDEX_TIP) continue;
      const p = project(i);
      ctx.beginPath();
      ctx.arc(p.x, p.y, 1.9, 0, TAU);
      ctx.fill();
    }

    ctx.restore();
  }

  private drawTrail(
    ctx: CanvasRenderingContext2D,
    mapper: FrameMapper,
    pointer: HandPointer,
    options: OverlayOptions,
  ): void {
    const trail = pointer.trail;
    if (trail.length < 3) return;

    ctx.save();
    ctx.lineCap = 'round';
    ctx.lineWidth = 1.5;

    for (let i = 1; i < trail.length; i += 1) {
      const t = i / (trail.length - 1);
      const from = mapper.boardToStage(trail[i - 1].x, trail[i - 1].y);
      const to = mapper.boardToStage(trail[i].x, trail[i].y);
      ctx.strokeStyle = withAlpha(options.trace, 0.22 * t * t);
      ctx.beginPath();
      ctx.moveTo(from.x, from.y);
      ctx.lineTo(to.x, to.y);
      ctx.stroke();
    }

    ctx.restore();
  }

  private drawNib(
    ctx: CanvasRenderingContext2D,
    mapper: FrameMapper,
    pointer: HandPointer,
    options: OverlayOptions,
  ): void {
    const p = mapper.boardToStage(pointer.tip.x, pointer.tip.y);
    const isEraser = options.tool === 'eraser';
    const accent = isEraser ? options.paper : options.color;
    const radius = Math.max((options.brush * mapper.pixelsPerUnit) / 2, 7);
    const down = pointer.isDown;

    ctx.save();
    ctx.lineCap = 'butt';

    // Ring — its radius is the actual stroke width, so you can size a line
    // before committing to it.
    ctx.beginPath();
    ctx.arc(p.x, p.y, radius, 0, TAU);
    if (down && !isEraser) {
      ctx.fillStyle = withAlpha(accent, 0.16);
      ctx.fill();
    }
    ctx.setLineDash(down ? [] : [3, 4]);
    ctx.strokeStyle = withAlpha(accent, down ? 0.95 : 0.5);
    ctx.lineWidth = down ? 1.75 : 1;
    ctx.stroke();
    ctx.setLineDash([]);

    // Confidence arc — how close the gesture is to engaging.
    if (!down && pointer.score > 0.02) {
      const sweep = clamp01(pointer.score) * TAU;
      ctx.beginPath();
      ctx.arc(p.x, p.y, radius + 6, -Math.PI / 2, -Math.PI / 2 + sweep);
      ctx.strokeStyle = withAlpha(accent, lerp(0.25, 0.9, clamp01(pointer.score)));
      ctx.lineWidth = 2;
      ctx.stroke();
    }

    // Registration ticks, retracted while the pen is down.
    const tickInner = down ? radius + 3 : radius + 10;
    const tickOuter = down ? radius + 7 : radius + 17;
    ctx.strokeStyle = withAlpha(down ? accent : options.dim, down ? 0.8 : 0.55);
    ctx.lineWidth = 1;
    ctx.beginPath();
    for (let i = 0; i < 4; i += 1) {
      const angle = (i * TAU) / 4;
      const cos = Math.cos(angle);
      const sin = Math.sin(angle);
      ctx.moveTo(p.x + cos * tickInner, p.y + sin * tickInner);
      ctx.lineTo(p.x + cos * tickOuter, p.y + sin * tickOuter);
    }
    ctx.stroke();

    // Centre: a filled point once inking, a hairline cross while hovering.
    if (down) {
      ctx.fillStyle = withAlpha(accent, 1);
      ctx.beginPath();
      ctx.arc(p.x, p.y, 2, 0, TAU);
      ctx.fill();
    } else {
      ctx.strokeStyle = withAlpha(accent, 0.8);
      ctx.beginPath();
      ctx.moveTo(p.x - 3.5, p.y);
      ctx.lineTo(p.x + 3.5, p.y);
      ctx.moveTo(p.x, p.y - 3.5);
      ctx.lineTo(p.x, p.y + 3.5);
      ctx.stroke();
    }

    this.drawCharge(ctx, pointer.id, p, radius, accent, options.now);
    ctx.restore();
  }

  /** One-shot ring that collapses into the nib the instant the pen engages. */
  private drawCharge(
    ctx: CanvasRenderingContext2D,
    id: number,
    p: { x: number; y: number },
    radius: number,
    accent: string,
    now: number,
  ): void {
    const startedAt = this.charges.get(id);
    if (startedAt === undefined) return;

    const t = (now - startedAt) / CHARGE_MS;
    if (t >= 1) {
      this.charges.delete(id);
      return;
    }

    const eased = 1 - (1 - t) ** 3;
    ctx.beginPath();
    ctx.arc(p.x, p.y, lerp(radius * 3, radius, eased), 0, TAU);
    ctx.strokeStyle = withAlpha(accent, (1 - t) * 0.9);
    ctx.lineWidth = 2;
    ctx.stroke();
  }

  reset(): void {
    this.charges.clear();
  }
}

File: drawone/render/overlay.py
"""Everything that is not ink: the tracked hand skeleton and the nib.

The nib is deliberately more than a dot. Air-drawing has no physical feedback, so
the cursor has to answer three questions at a glance: where the pen is, how thick
the line will be (the ring is the true stroke width), and how close the gesture is
to engaging (the arc). Without that last one, a gesture that nearly fires just
feels broken.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Dict, Sequence

from ..config import THEME
from ..lib.maths import clamp01, lerp
from ..tracking.frame_mapper import FrameMapper
from ..tracking.landmarks import FINGER_BONES, LM, PALM_BONES
from ..tracking.pointers import HandPointer
from ..types import Tool
from . import canvas
from .canvas import Layer, Point
from .color import Bgr, to_bgr

CHARGE_MS = 170.0
OTHER_FINGERS = ("thumb", "middle", "ring", "pinky")


@dataclass(frozen=True)
class OverlayOptions:
    show_skeleton: bool
    tool: Tool
    color: str
    #: Current brush size in board units.
    brush: float
    #: Monotonic milliseconds, for time-based cues.
    now: float
    trace: str = THEME["trace"]
    dim: str = THEME["dim"]
    paper: str = THEME["paper"]


class OverlayRenderer:
    def __init__(self) -> None:
        self._charges: Dict[int, float] = {}

    def render(
        self,
        layer: Layer,
        mapper: FrameMapper,
        pointers: Sequence[HandPointer],
        options: OverlayOptions,
    ) -> None:
        layer.clear()
        if not mapper.ready:
            return

        for pointer in pointers:
            if pointer.just_pressed:
                self._charges[pointer.id] = options.now
            if options.show_skeleton:
                self._draw_skeleton(layer, mapper, pointer, options)
            if not pointer.is_down:
                self._draw_trail(layer, mapper, pointer, options)

        # Nibs last so they sit above every skeleton when hands overlap.
        for pointer in pointers:
            self._draw_nib(layer, mapper, pointer, options)

    def _draw_skeleton(
        self,
        layer: Layer,
        mapper: FrameMapper,
        pointer: HandPointer,
        options: OverlayOptions,
    ) -> None:
        landmarks = pointer.landmarks

        def project(index: int) -> Point:
            lm = landmarks[index]
            p = mapper.normalized_to_stage(lm.x, lm.y)
            return (p.x, p.y)

        # Step back while drawing so the skeleton never competes with the stroke.
        fade = 0.45 if pointer.is_down else 1.0
        trace = to_bgr(options.trace)

        def bones(pairs, alpha: float, width: float) -> None:
            for a, b in pairs:
                canvas.line(layer, project(a), project(b), trace, alpha * fade, width)

        bones(PALM_BONES, 0.28, 1.25)
        for finger in OTHER_FINGERS:
            bones(FINGER_BONES[finger], 0.28, 1.25)
        # The index chain is the one that matters — it carries the nib.
        bones(FINGER_BONES["index"], 0.85, 1.75)

        for i in range(len(landmarks)):
            if i == LM.INDEX_TIP:
                continue
            canvas.circle(layer, project(i), 1.9, trace, 0.5 * fade, fill=True)

    def _draw_trail(
        self,
        layer: Layer,
        mapper: FrameMapper,
        pointer: HandPointer,
        options: OverlayOptions,
    ) -> None:
        trail = pointer.trail
        if len(trail) < 3:
            return

        trace = to_bgr(options.trace)
        for i in range(1, len(trail)):
            t = i / (len(trail) - 1)
            start = mapper.board_to_stage(trail[i - 1].x, trail[i - 1].y)
            end = mapper.board_to_stage(trail[i].x, trail[i].y)
            canvas.line(
                layer, (start.x, start.y), (end.x, end.y), trace, 0.22 * t * t, 1.5
            )

    def _draw_nib(
        self,
        layer: Layer,
        mapper: FrameMapper,
        pointer: HandPointer,
        options: OverlayOptions,
    ) -> None:
        point = mapper.board_to_stage(pointer.tip.x, pointer.tip.y)
        p: Point = (point.x, point.y)
        is_eraser = options.tool == "eraser"
        accent = to_bgr(options.paper if is_eraser else options.color)
        radius = max((options.brush * mapper.pixels_per_unit) / 2, 7.0)
        down = pointer.is_down

        # Ring — its radius is the actual stroke width, so you can size a line
        # before committing to it.
        if down and not is_eraser:
            canvas.circle(layer, p, radius, accent, 0.16, fill=True)
        if down:
            canvas.circle(layer, p, radius, accent, 0.95, width=1.75)
        else:
            canvas.dashed_circle(layer, p, radius, accent, 0.5, width=1.0)

        # Confidence arc — how close the gesture is to engaging.
        if not down and pointer.score > 0.02:
            sweep = clamp01(pointer.score) * 360.0
            canvas.arc(
                layer,
                p,
                radius + 6,
                -90.0,
                -90.0 + sweep,
                accent,
                lerp(0.25, 0.9, clamp01(pointer.score)),
                2.0,
            )

        # Registration ticks, retracted while the pen is down.
        tick_inner = radius + 3 if down else radius + 10
        tick_outer = radius + 7 if down else radius + 17
        tick_color = accent if down else to_bgr(options.dim)
        tick_alpha = 0.8 if down else 0.55
        for i in range(4):
            angle = i * math.pi / 2
            cos = math.cos(angle)
            sin = math.sin(angle)
            canvas.line(
                layer,
                (p[0] + cos * tick_inner, p[1] + sin * tick_inner),
                (p[0] + cos * tick_outer, p[1] + sin * tick_outer),
                tick_color,
                tick_alpha,
                1.0,
            )

        # Centre: a filled point once inking, a hairline cross while hovering.
        if down:
            canvas.circle(layer, p, 2.0, accent, 1.0, fill=True)
        else:
            canvas.line(layer, (p[0] - 3.5, p[1]), (p[0] + 3.5, p[1]), accent, 0.8, 1.0)
            canvas.line(layer, (p[0], p[1] - 3.5), (p[0], p[1] + 3.5), accent, 0.8, 1.0)

        self._draw_charge(layer, pointer.id, p, radius, accent, options.now)

    def _draw_charge(
        self,
        layer: Layer,
        pointer_id: int,
        p: Point,
        radius: float,
        accent: Bgr,
        now: float,
    ) -> None:
        """One-shot ring that collapses into the nib the instant the pen engages."""
        started_at = self._charges.get(pointer_id)
        if started_at is None:
            return

        t = (now - started_at) / CHARGE_MS
        if t >= 1:
            del self._charges[pointer_id]
            return

        eased = 1 - (1 - t) ** 3
        canvas.circle(
            layer, p, lerp(radius * 3, radius, eased), accent, (1 - t) * 0.9, width=2.0
        )

    def reset(self) -> None:
        self._charges.clear()
